File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/TabListSingleAction.java
/*
 * Copyright (C) 2017 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.action.item;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Parcel;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Spinner;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.utils.Logger;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;

public class TabListSingleAction extends SingleAction implements Parcelable {
    public static final int MODE_NORMAL = 0;
    public static final int MODE_REVERSE = 1;
    public static final int MODE_HORIZONTAL = 2;

    private static final String TAG = "TabListSingleAction";
    private static final String FIELD_NAME_REVERSE = "0";
    private static final String FIELD_NAME_MODE = "1";
    private static final String FIELD_NAME_LEFT_BUTTON = "2";
    private int mode = MODE_HORIZONTAL;
    private boolean leftButton;

    public TabListSingleAction(int id, JsonParser parser) throws IOException {
        super(id);

        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (FIELD_NAME_REVERSE.equals(parser.getCurrentName())) {
                    switch (parser.nextToken()) {
                        case VALUE_TRUE:
                            mode = MODE_REVERSE;
                            break;
                        case VALUE_FALSE:
                            mode = MODE_NORMAL;
                            break;
                        default:
                            Logger.w(TAG, "current token is not boolean value : " + parser.getCurrentToken().toString());
                            break;
                    }
                } else if (FIELD_NAME_MODE.equals(parser.getCurrentName())) {
                    if (parser.nextValue().isNumeric())
                        mode = parser.getIntValue();
                } else if (FIELD_NAME_LEFT_BUTTON.equals(parser.getCurrentName())) {
                    if (parser.nextValue().isBoolean())
                        leftButton = parser.getBooleanValue();
                } else {
                    parser.skipChildren();
                }

            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        generator.writeNumberField(FIELD_NAME_MODE, mode);
        generator.writeBooleanField(FIELD_NAME_LEFT_BUTTON, leftButton);
        generator.writeEndObject();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeInt(mode);
    }

    protected TabListSingleAction(Parcel source) {
        super(source.readInt());
        mode = source.readInt();
    }

    public static final Creator<TabListSingleAction> CREATOR = new Creator<TabListSingleAction>() {
        @Override
        public TabListSingleAction createFromParcel(Parcel source) {
            return new TabListSingleAction(source);
        }

        @Override
        public TabListSingleAction[] newArray(int size) {
            return new TabListSingleAction[size];
        }
    };

    @Override
    public StartActivityInfo showSubPreference(ActionActivity context) {
        View view = LayoutInflater.from(context).inflate(R.layout.action_tab_list, null);
        final Spinner modeSpinner = view.findViewById(R.id.modeSpinner);
        final Spinner buttonSpinner = view.findViewById(R.id.buttonSpinner);

        modeSpinner.setSelection(mode);
        buttonSpinner.setSelection(leftButton ? 1 : 0);
        new AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mode = modeSpinner.getSelectedItemPosition();
                        leftButton = buttonSpinner.getSelectedItemPosition() > 0;
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();

        return null;
    }

    public int getMode() {
        return mode;
    }

    public boolean isLeftButton() {
        return leftButton;
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/item/TabListSingleAction.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.item

import android.app.AlertDialog
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import android.widget.Spinner
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import jp.hazuki.yuzubrowser.legacy.action.view.ActionActivity
import jp.hazuki.yuzubrowser.ui.app.StartActivityInfo
import java.io.IOException

class TabListSingleAction : SingleAction, Parcelable {
    var mode = MODE_HORIZONTAL
        private set
    var isLeftButton: Boolean = false
        private set
    var lastTabMode = 0
        private set

    @Throws(IOException::class)
    constructor(id: Int, reader: JsonReader?) : super(id) {
        if (reader != null) {
            if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.peek() != JsonReader.Token.NAME) return
                when (reader.nextName()) {
                    FIELD_NAME_REVERSE -> {
                        if (reader.peek() != JsonReader.Token.BOOLEAN) return
                        mode = if (reader.nextBoolean()) MODE_REVERSE else MODE_NORMAL
                    }
                    FIELD_NAME_MODE -> {
                        if (reader.peek() != JsonReader.Token.NUMBER) return
                        mode = reader.nextInt()
                    }
                    FIELD_NAME_LEFT_BUTTON -> {
                        if (reader.peek() != JsonReader.Token.BOOLEAN) return
                        isLeftButton = reader.nextBoolean()
                    }
                    FIELD_NAME_LAST_TAB_MODE -> {
                        if (reader.peek() != JsonReader.Token.NUMBER) return
                        lastTabMode = reader.nextInt()
                    }
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(writer: JsonWriter) {
        writer.value(id)
        writer.beginObject()
        writer.name(FIELD_NAME_MODE)
        writer.value(mode)
        writer.name(FIELD_NAME_LEFT_BUTTON)
        writer.value(isLeftButton)
        writer.name(FIELD_NAME_LAST_TAB_MODE)
        writer.value(lastTabMode)
        writer.endObject()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeInt(mode)
        dest.writeInt(lastTabMode)
    }

    private constructor(source: Parcel) : super(source.readInt()) {
        mode = source.readInt()
        lastTabMode = source.readInt()
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        val view = View.inflate(context, R.layout.action_tab_list, null)
        val modeSpinner = view.findViewById<Spinner>(R.id.modeSpinner)
        val buttonSpinner = view.findViewById<Spinner>(R.id.buttonSpinner)
        val lastTabSpinner = view.findViewById<Spinner>(R.id.lastTabSpinner)

        modeSpinner.setSelection(mode)
        buttonSpinner.setSelection(if (isLeftButton) 1 else 0)
        lastTabSpinner.setSelection(lastTabMode)
        AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    mode = modeSpinner.selectedItemPosition
                    isLeftButton = buttonSpinner.selectedItemPosition > 0
                    lastTabMode = lastTabSpinner.selectedItemPosition
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()

        return null
    }

    companion object {
        const val MODE_NORMAL = 0
        const val MODE_REVERSE = 1
        const val MODE_HORIZONTAL = 2

        const val LAST_TAB_MODE_NONE = 0
        const val LAST_TAB_MODE_NEW_TAB = 1
        const val LAST_TAB_MODE_FINISH = 2
        const val LAST_TAB_MODE_FINISH_WITH_ALERT = 3

        private const val TAG = "TabListSingleAction"
        private const val FIELD_NAME_REVERSE = "0"
        private const val FIELD_NAME_MODE = "1"
        private const val FIELD_NAME_LEFT_BUTTON = "2"
        private const val FIELD_NAME_LAST_TAB_MODE = "3"

        @JvmField
        val CREATOR: Parcelable.Creator<TabListSingleAction> = object : Parcelable.Creator<TabListSingleAction> {
            override fun createFromParcel(source: Parcel): TabListSingleAction {
                return TabListSingleAction(source)
            }

            override fun newArray(size: Int): Array<TabListSingleAction?> {
                return arrayOfNulls(size)
            }
        }
    }
}
